open_instruct/grpo_fast.py

```diff
@@ -1963,67 +1975,73 @@ def one_training_step(
-                    [
-                        policy_group.models[i].save_model.remote(step_dir, chat_template_name, tokenizer)
-                        for i in range(args.world_size)
-                    ],
-                    desc=f"Saving model at step {training_step}",
-                )
-                if args.try_launch_beaker_eval_jobs_on_weka and is_beaker_job():
-                    leaderboard_name = f"{args.hf_repo_revision}_step_{training_step}"
-                    for i in range(args.world_size):
-                        policy_group.models[i].launch_ai2_evals_on_weka_wrapper.remote(
-                            step_dir, leaderboard_name, wandb_url, training_step
-                        )
-        if (
-            args.checkpoint_state_freq > 0
-            and training_step % args.checkpoint_state_freq == 0
-            and args.checkpoint_state_dir is not None
-        ):
-            with Timer("[Main Thread] 🗡️ Saving checkpoint state"):
-                client_state = {"training_step": training_step}
-                ray_get_with_progress(
-                    [
-                        policy_group.models[i].save_checkpoint_state.remote(args.checkpoint_state_dir, client_state)
-                        for i in range(args.world_size)
-                    ],
-                    desc=f"Saving checkpoint state at step {training_step}",
-                )
-                logger.info(f"Saved checkpoint state at step {training_step} to {args.checkpoint_state_dir}")
+    if (
+        args.checkpoint_state_freq > 0
+        and training_step % args.checkpoint_state_freq == 0
+        and args.checkpoint_state_dir is not None
+    ):
+        with Timer("[Main Thread] 🗡️ Saving checkpoint state") as timer:
+            client_state = {"training_step": training_step}
+            ray_get_with_progress(
+                [
+                    policy_group.models[i].save_checkpoint_state.remote(args.checkpoint_state_dir, client_state)
+                    for i in range(args.world_size)
+                ],
+                desc=f"Saving checkpoint state at step {training_step}",
+            )
+            logger.info(f"Saved checkpoint state at step {training_step} to {args.checkpoint_state_dir}")
+        save_time += timer.duration
 
     if len(update_ref_policy_future) > 0:
         with Timer("[Main Thread] 🔃 Updating reference policy"):
             ray_get_with_progress(update_ref_policy_future, desc="Updating reference policy")
 
-    return average_metrics
+    average_metrics = {k: sum(m[k] for m in metrics_list) / len(metrics_list) for k in metrics_list[0]}
+    total_time = time.time() - start_time
+    metrics = {
+        "episode": episode,
+        "training_step": training_step,
+        "val/num_total_tokens": num_total_tokens,
+        "epoch": episode / args.num_samples_per_prompt_rollout / len(train_dataset),
+        "tokens_per_second": num_total_tokens / total_time,
```

Comment:
@finbarrtimbers we already have tokens_per_second, but I can put it under `time/` 
how can I add MFU?

Reply:
does this account for the generation time as well as training time?

Reply:
This is the time for a single training step. 

For inference, you can use num_params * bytes_per_param ([reference](https://kipp.ly/transformer-inference-arithmetic/)).

Reply:
prime-rl does  `flop_per_token = 6 * num_params + 12 * l * h * q * t` instead of `3*num_params * bytes per param` which I assume is 6 because 2 bytes per bf16 and then 12* is the generation. Wyt?

https://github.com/PrimeIntellect-ai/prime-rl/blob/9657a591added96d2671f32f9307bd4d5d92939c/src/prime_rl/trainer/perf.py#L75